### PR TITLE
ResumableJS fault tolerant attachment upload

### PR DESCRIPTION
**Pull request by henrique-borba**
This PR aims to redo the attachment upload methods using ResumableJS.

## Screen

Screenshot_20211023_220605

Screenshot_20211026_110858

## Features

- Fault Tolerant upload that allows you to pause or resume an upload that failed due to network issues
- Due to the chunky nature of ResumableJS, very large files can be uploaded without breaking HTML5 Upload API limits or HTTP body size limits sent to the server.
- Now we have a progress bar
- If the user leaves the message composition screen and returns later, the upload will continue from the last piece of the partially uploaded file, preventing the loss of upload progress.
- This update will fix most of the current issues related to attachments and draft attachments
- The administrator will be able to clean up the partial chunks of unused files, this prevents storage limit issues and excessive garbage growth

## Important

- Drafts will not be automatically saved after uploading or removing a file. The user must ALWAYS press the save button to save their changes after send/remove attachments (this can be changed easily, but I found it more intuitive)
- Session Cache will no longer be used to manage attachments

## Core Updates

- This update allows the hm_ajax_hook to be sent via a GET parameter

### CODE REVIEW

File: modules/smtp/modules.php [hunk 1]
```diff
@@ -173,68 +160,61 @@ public function process() {
...
-        if ($success) {
-            delete_uploaded_files($this->session, $form['draft_id']);
-            if (delete_draft($form['draft_id'], $this->session)) {
-                Hm_Msgs::add('Draft deleted');
-                $this->out('draft_id', $form['draft_id']);
+        $imap_profile = find_imap_by_smtp(
+            $this->user_config->get('imap_servers'),
+            $this->user_config->get('smtp_servers')[$this->request->get['draft_smtp']]
+        );
+        $from = $this->request->get['draft_smtp'];
```
[the remark this answers is not in the document]
Comment (henrique-borba) [reply modules/smtp/modules.php:190]: Yes, it is all working fine with the other parameters

File: modules/smtp/modules.php [hunk 2]
```diff
@@ -173,68 +160,61 @@ public function process() {
...
-class Hm_Handler_process_delete_draft extends Hm_Handler_Module {
+class Hm_Handler_upload_chunk extends Hm_Handler_Module {
     public function process() {
-        list($success, $form) = $this->process_form(array('draft_id'));
-        if ($success) {
-            delete_uploaded_files($this->session, $form['draft_id']);
-            if (delete_draft($form['draft_id'], $this->session)) {
-                Hm_Msgs::add('Draft deleted');
-                $this->out('draft_id', $form['draft_id']);
+        $imap_profile = find_imap_by_smtp(
```
[the remark this answers is not in the document]
Comment (henrique-borba) [reply modules/smtp/modules.php]: Code garbage, `$imap_profile` is not needed in `Hm_Handler_upload_chunk`

**Comment by jasonmunro** [comment]
Just gave this a try and on upload I get an error and the progress bar never goes away and the "pause" link remains.
```[10-Jan-2022 18:42:31 Africa/Abidjan] PHP Fatal error:  Uncaught Error: Class 'Hm_IMAP_List' not found in /home/jason/cypht/modules/profiles/hm-profiles.php:89
Stack trace:
#0 /home/jason/cypht/modules/profiles/hm-profiles.php(25): Hm_Profiles->load_legacy()
#1 /home/jason/cypht/modules/profiles/hm-profiles.php(19): Hm_Profiles->load()
#2 /home/jason/cypht/modules/profiles/modules.php(41): Hm_Profiles->__construct()
#3 /home/jason/cypht/lib/modules_exec.php(175): Hm_Handler_compose_profile_data->process()
#4 /home/jason/cypht/lib/modules_exec.php(153): Hm_Module_Exec->run_handler_module()
#5 /home/jason/cypht/lib/dispatch.php(220): Hm_Module_Exec->run_handler_modules()
#6 /home/jason/cypht/lib/dispatch.php(190): Hm_Dispatch->process_request()
#7 /home/jason/cypht/index.php(50): Hm_Dispatch->__construct()
#8 {main}
  thrown in /home/jason/cypht/modules/profiles/hm-profiles.php on line 89
[10-Jan-2022 18:42:31 Africa/Abidjan] PHP Fatal error:  Uncaught Error: Class 'Hm_IMAP_List' not found in /home/jason/cypht/modules/profiles/hm-profiles.php:89
Stack trace:
#0 /home/jason/cypht/modules/profiles/hm-profiles.php(25): Hm_Profiles->load_legacy()
#1 /home/jason/cypht/modules/profiles/hm-profiles.php(19): Hm_Profiles->load()
#2 /home/jason/cypht/modules/profiles/modules.php(41): Hm_Profiles->__construct()
#3 /home/jason/cypht/lib/modules_exec.php(175): Hm_Handler_compose_profile_data->process()
#4 /home/jason/cypht/lib/modules_exec.php(153): Hm_Module_Exec->run_handler_module()
#5 /home/jason/cypht/lib/dispatch.php(220): Hm_Module_Exec->run_handler_modules()
#6 /home/jason/cypht/lib/dispatch.php(190): Hm_Dispatch->process_request()
#7 /home/jason/cypht/index.php(50): Hm_Dispatch->__construct()
#8 {main}
  thrown in /home/jason/cypht/modules/profiles/hm-profiles.php on line 89
[10-Jan-2022 18:42:31 Africa/Abidjan] PHP Fatal error:  Uncaught Error: Class 'Hm_IMAP_List' not found in /home/jason/cypht/modules/profiles/hm-profiles.php:89
Stack trace:
#0 /home/jason/cypht/modules/profiles/hm-profiles.php(25): Hm_Profiles->load_legacy()
#1 /home/jason/cypht/modules/profiles/hm-profiles.php(19): Hm_Profiles->load()
#2 /home/jason/cypht/modules/profiles/modules.php(41): Hm_Profiles->__construct()
#3 /home/jason/cypht/lib/modules_exec.php(175): Hm_Handler_compose_profile_data->process()
#4 /home/jason/cypht/lib/modules_exec.php(153): Hm_Module_Exec->run_handler_module()
#5 /home/jason/cypht/lib/dispatch.php(220): Hm_Module_Exec->run_handler_modules()
#6 /home/jason/cypht/lib/dispatch.php(190): Hm_Dispatch->process_request()
#7 /home/jason/cypht/index.php(50): Hm_Dispatch->__construct()
#8 {main}
  thrown in /home/jason/cypht/modules/profiles/hm-profiles.php on line 89```

**Comment by jasonmunro** [comment]
Just tried the admin "clear chunks" feature and because my attachments were in the top level of my cypht config it just blew away all of my app data (like my sqlite db with all my settings). Seems like this is a bit of a problem :)

**Comment by jasonmunro** [comment]
> Just gave this a try and on upload I get an error and the progress bar never goes away and the "pause" link remains.
> 
> ```
> Stack trace:
> #0 /home/jason/cypht/modules/profiles/hm-profiles.php(25): Hm_Profiles->load_legacy()
> #1 /home/jason/cypht/modules/profiles/hm-profiles.php(19): Hm_Profiles->load()
> #2 /home/jason/cypht/modules/profiles/modules.php(41): Hm_Profiles->__construct()
> #3 /home/jason/cypht/lib/modules_exec.php(175): Hm_Handler_compose_profile_data->process()
> #4 /home/jason/cypht/lib/modules_exec.php(153): Hm_Module_Exec->run_handler_module()
> #5 /home/jason/cypht/lib/dispatch.php(220): Hm_Module_Exec->run_handler_modules()
> #6 /home/jason/cypht/lib/dispatch.php(190): Hm_Dispatch->process_request()
> #7 /home/jason/cypht/index.php(50): Hm_Dispatch->__construct()
> #8 {main}
>   thrown in /home/jason/cypht/modules/profiles/hm-profiles.php on line 89
> [10-Jan-2022 18:42:31 Africa/Abidjan] PHP Fatal error:  Uncaught Error: Class 'Hm_IMAP_List' not found in /home/jason/cypht/modules/profiles/hm-profiles.php:89
> Stack trace:
> #0 /home/jason/cypht/modules/profiles/hm-profiles.php(25): Hm_Profiles->load_legacy()
> #1 /home/jason/cypht/modules/profiles/hm-profiles.php(19): Hm_Profiles->load()
> #2 /home/jason/cypht/modules/profiles/modules.php(41): Hm_Profiles->__construct()
> #3 /home/jason/cypht/lib/modules_exec.php(175): Hm_Handler_compose_profile_data->process()
> #4 /home/jason/cypht/lib/modules_exec.php(153): Hm_Module_Exec->run_handler_module()
> #5 /home/jason/cypht/lib/dispatch.php(220): Hm_Module_Exec->run_handler_modules()
> #6 /home/jason/cypht/lib/dispatch.php(190): Hm_Dispatch->process_request()
> #7 /home/jason/cypht/index.php(50): Hm_Dispatch->__construct()
> #8 {main}
>   thrown in /home/jason/cypht/modules/profiles/hm-profiles.php on line 89
> [10-Jan-2022 18:42:31 Africa/Abidjan] PHP Fatal error:  Uncaught Error: Class 'Hm_IMAP_List' not found in /home/jason/cypht/modules/profiles/hm-profiles.php:89
> Stack trace:
> #0 /home/jason/cypht/modules/profiles/hm-profiles.php(25): Hm_Profiles->load_legacy()
> #1 /home/jason/cypht/modules/profiles/hm-profiles.php(19): Hm_Profiles->load()
> #2 /home/jason/cypht/modules/profiles/modules.php(41): Hm_Profiles->__construct()
> #3 /home/jason/cypht/lib/modules_exec.php(175): Hm_Handler_compose_profile_data->process()
> #4 /home/jason/cypht/lib/modules_exec.php(153): Hm_Module_Exec->run_handler_module()
> #5 /home/jason/cypht/lib/dispatch.php(220): Hm_Module_Exec->run_handler_modules()
> #6 /home/jason/cypht/lib/dispatch.php(190): Hm_Dispatch->process_request()
> #7 /home/jason/cypht/index.php(50): Hm_Dispatch->__construct()
> #8 {main}
>   thrown in /home/jason/cypht/modules/profiles/hm-profiles.php on line 89```
> ```

Looks like this is specific to not having a configured profile. We should still be able to attache/send mail without a profile.

**Comment by henrique-borba** [comment]
I will reset my environment to try to explain the problem, if the module is not able to find the hm_IMAP class then nothing will work.

About the removed files, I should have used the same **if** (https://github.com/jasonmunro/cypht/blob/1397ec625aaa1d0b807cb2932dd6bd9e327db80d/modules/smtp/modules.php#L729) that calculates the size of the chunks in the removal loop

**Comment by jasonmunro** [comment]
> I will reset my environment to try to replicate the problem, if the module is not able to find the hm_IMAP class then nothing will work.

Yep I get that, however in this case I was attempting to just attach a file without a profile enabled for the selected SMTP server and it fails.

**Comment by henrique-borba** [comment]
> > I will reset my environment to try to replicate the problem, if the module is not able to find the hm_IMAP class then nothing will work.
> 
> Yep I get that, however in this case I was attempting to just attach a file without a profile enabled for the selected SMTP server and it fails.

I was almost blowing my head here without being able to replicate, all my tests were with Profiles enabled lol

**Comment by henrique-borba** [comment]
@jasonmunro @marclaporte I was able to reproduce and now it is fixed

**Comment by jasonmunro** [comment]
> @jasonmunro @marclaporte I was able to reproduce and now it is fixed

Yep looks good. I wonder if the progress bar can disappear after the upload is complete? WIth it just staying there fully green it makes me think the upload is still happening. One final issue is when I send a message the attachment is in the E-mail but does not show up due to a missing header value. Here is the MIME header for the attachment block:

```--a55OrHY6y43Lm4bDbEmqEA7nFm5mdYa1h95AHAxk9q3X5EkKf9HZW6mAZBGVzEO
Content-Type: ; name="multipageproject.pdf"
Content-Description: multipageproject.pdf
Content-Disposition: attachment; filename="multipageproject.pdf"
Content-Transfer-Encoding: base64
```

Looks like the initial content type is missing for some reason. Pretty sure if that is correctly populated the attachment will show up on the delivered mail.

**Comment by jasonmunro** [comment]
One more issue - attachment storage on the server is not user specific at all, so I can easily overwrite another users attachments with the same name (!) (tested this with 2 users and the same attached file, both went into a subdir "0" but the filename is the original and the second user overwrote the first)

**Comment by dumblob** [comment]
> One more issue - attachment storage on the server is not user specific at all, so I can easily overwrite another users attachments with the same name (!)

Actually, I'd think the storage must not just be user-specific, not even email-specific, not even folder-specific etc., not even file-specific (i.e. 1:1 `memcmp()`), not even time-specific (because of time fluctuation on the server), but I'm pretty sure it has to be instance-specific - i.e. every attempt to upload has to lead to **unique** file on the server storage (while preserving the metadata - especially the original file name - somewhere else - at least in the email itself :wink:).

Of course, in case it's a continuation of a broken session (either manually suspended or due to e.g. network issue), then it shall automatically continue (in case the browser tab didn't close and thus the session details - especially the permission of the user given to the browser to upload the given file is granted) or ask the user whether to continue (in which case a select file dialog would pop up and then a quick check - e.g. the first and last already uploaded byte of the already uploaded part on the server - would be compared "just in case" the user selected a wrong file to continue with to save a trouble of waiting minutes for large upload over a slow mobile connection - which is exactly the use case of this PR - just to realize the hash of the final file on the server doesn't match the one the user has chosen) or whether to postpone the decision of continuation or whether to cancel the whole upload altogether.

**Comment by henrique-borba** [comment]
> > @jasonmunro @marclaporte I was able to reproduce and now it is fixed
> 
> Yep looks good. I wonder if the progress bar can disappear after the upload is complete? WIth it just staying there fully green it makes me think the upload is still happening. One final issue is when I send a message the attachment is in the E-mail but does not show up due to a missing header value. Here is the MIME header for the attachment block:
> 
> ```
> Content-Type: ; name="multipageproject.pdf"
> Content-Description: multipageproject.pdf
> Content-Disposition: attachment; filename="multipageproject.pdf"
> Content-Transfer-Encoding: base64
> ```
> 
> Looks like the initial content type is missing for some reason. Pretty sure if that is correctly populated the attachment will show up on the delivered mail.

Sure, a fadeOut will do the job. I was able to send and download my attachments normally between clients =/. I think you are not receiving the attachments because of the folder issue below

**Comment by henrique-borba** [comment]
> One more issue - attachment storage on the server is not user specific at all, so I can easily overwrite another users attachments with the same name (!) (tested this with 2 users and the same attached file, both went into a subdir "0" but the filename is the original and the second user overwrote the first)

I think I know what's going on, I have two accounts and they have two different smtp servers, so I never had them both using the same folder (the folder is the smtp_id). I guess in your case you have the same smtp_id for both accounts?

**Comment by henrique-borba** [comment]
> > One more issue - attachment storage on the server is not user specific at all, so I can easily overwrite another users attachments with the same name (!)
> 
> Actually, I'd think the storage must not just be user-specific, not even email-specific, not even folder-specific etc., not even file-specific (i.e. 1:1 `memcmp()`), not even time-specific (because of time fluctuation on the server), but I'm pretty sure it has to be instance-specific - i.e. every attempt to upload has to lead to **unique** file on the server storage (while preserving the metadata - especially the original file name - somewhere else - at least in the email itself 😉).
> 
> Of course, in case it's a continuation of a broken session (either manually suspended or due to e.g. network issue), then it shall automatically continue (in case the browser tab didn't close and thus the session details - especially the permission of the user given to the browser to upload the given file is granted) or ask the user whether to continue (in which case a select file dialog would pop up and then a quick check - e.g. the first and last already uploaded byte of the already uploaded part on the server - would be compared "just in case" the user selected a wrong file to continue with to save a trouble of waiting minutes for large upload over a slow mobile connection - which is exactly the use case of this PR - just to realize the hash of the final file on the server doesn't match the one the user has chosen) or whether to postpone the decision of continuation or whether to cancel the whole upload altogether.

I agree, we could hash the filename using the 2-way hash Cypht already use to encrypt the file content. But the same file must match the one stored as this is the way ResumableJS can find a partial uploaded file to continue the upload.

**Comment by jasonmunro** [comment]
> > One more issue - attachment storage on the server is not user specific at all, so I can easily overwrite another users attachments with the same name (!) (tested this with 2 users and the same attached file, both went into a subdir "0" but the filename is the original and the second user overwrote the first)
> 
> I think I know what's going on, I have two accounts and they have two different smtp servers, so I never had them both using the same folder (the folder is the smtp_id). I guess in your case you have the same smtp_id for both accounts?

Yep, SMTP server ids are just incrementing integers and not unique across users

**Comment by jasonmunro** [comment]
> > > @jasonmunro @marclaporte I was able to reproduce and now it is fixed
> > 
> > 
> > Yep looks good. I wonder if the progress bar can disappear after the upload is complete? WIth it just staying there fully green it makes me think the upload is still happening. One final issue is when I send a message the attachment is in the E-mail but does not show up due to a missing header value. Here is the MIME header for the attachment block:
> > ```
> > Content-Type: ; name="multipageproject.pdf"
> > Content-Description: multipageproject.pdf
> > Content-Disposition: attachment; filename="multipageproject.pdf"
> > Content-Transfer-Encoding: base64
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Looks like the initial content type is missing for some reason. Pretty sure if that is correctly populated the attachment will show up on the delivered mail.
> 
> Sure, a fadeOut will do the job. I was able to send and download my attachments normally between clients =/. I think you are not receiving the attachments because of the folder issue below

To clarify, the attachment is present but not being parsed as such by the IMAP server so either Dovecot or Cypht is failing somehow, but we should not be sending out messages without a Content-type value of at least application/octetstream  and preferably the correct MIME type for the attachment so Cypht knows how to present it

**Comment by jasonmunro** [comment]
To further clarify, this was in my received message headers for the attachment block:
```
 Content-Type: ; name="multipageproject.pdf"
 ```
 Should be:
 ```
 Content-Type: application/pdf; name="multipageproject.pdf"

**Comment by henrique-borba** [comment]
@jasonmunro I sent the correction of the Content-Type of the header and the new name of the directories that now uses the username.

I'm going to leave the hash of the filenames in a separate PR because I still need to come up with a solution.

**Comment by jasonmunro** [comment]
Looks good here, thanks for following up!

**Pull request merged**